src/hotspot/share/gc/g1/g1RemSet.cpp

```diff
@@ -1188,44 +1197,23 @@ class G1MergeHeapRootsTask : public AbstractGangTask {
       mark_card(to_process);
     }
 
-    void next_coarse_prt(uint const region_idx) {
-      if (!remember_if_interesting(region_idx)) {
-        return;
-      }
-
-      _merged_coarse++;
-
-      start_iterate(region_idx);
-      _cards_dirty += _ct->mark_region_dirty(_region_base_idx, HeapRegion::CardsPerRegion);
-      _scan_state->set_chunk_region_dirty(_region_base_idx);
-    }
-
-    void next_fine_prt(uint const region_idx, BitMap* bm) {
-      if (!remember_if_interesting(region_idx)) {
-        return;
-      }
-
-      _merged_fine++;
-
-      start_iterate(region_idx);
-      BitMap::idx_t cur = bm->get_next_one_offset(0);
-      while (cur != bm->size()) {
-        do_card((uint)cur);
-        cur = bm->get_next_one_offset(cur + 1);
+    // Returns whether the given region actually needs iteration.
+    bool start_iterate(uint const tag, uint const region_idx) {
+      assert(tag < G1GCPhaseTimes::MergeRSDirtyCards, "invalid tag %u", tag);
+      if (remember_if_interesting(region_idx)) {
+        _region_base_idx = (size_t)region_idx << HeapRegion::LogCardsPerRegion;
+        _stats.inc_card_set_merged(tag);
+        return true;
```

Comment:
Not entierly sure where to put this comment, but this closure is used both as a heap-region closure and as a closure for container-iterations. I think this should be documented somehow. Maybe both above this closure and maybe also list what closures are used for container-iteration.

Reply:
There is already a small comment that reads
```
// Visitor for remembered sets, dropping entries onto the card table.
```
at the top of the class declaration. I'll expand on that.